physics/discrete_trajectory_segment.hpp

```diff
@@ -136,6 +136,12 @@ class DiscreteTrajectorySegment : public Trajectory<Frame> {
       DiscreteTrajectorySegmentIterator<Frame> self);
 
  private:
+  // Versions of find, lower_bound, and upper_bound that use optionals to
+  // indicate absense. Useful for bypassing expensive end() call.
+  std::optional<iterator> FindImpl(Instant const& t) const;
```

Comment:
I don't think that we love the suffix `Impl`, we don't use that a lot.  Furthermore, it's not so clear at the call site how the `Impl` differ from the public method.  I'd call these `FindOrNullopt`, etc. which will make the role of these methods more explicit.

Reply:
Done.